src/projects.h

```diff
@@ -165,168 +167,320 @@ typedef struct { double u, v; } projUV;
-typedef struct { double lam, phi; } LP;
-typedef struct { double x, y, z; } XYZ;
+typedef struct { double x, y; }        XY;
+typedef struct { double x, y, z; }     XYZ;
+typedef struct { double lam, phi; }    LP;
 typedef struct { double lam, phi, z; } LPZ;
-#endif
+typedef struct { double u, v; }        UV;
+typedef struct { double u, v, w; }     UVW;
+#endif  /* ndef PJ_LIB__ */
+#endif  /* ndef PROJ_H   */
 
-typedef union { double  f; int  i; char *s; } PROJVALUE;
+
+/* Forward declarations and typedefs for stuff needed inside the PJ object */
 struct PJconsts;
+struct pj_opaque;
+struct ARG_list;
+struct FACTORS;
+struct PJ_REGION_S;
+typedef struct PJ_REGION_S  PJ_Region;
+typedef struct ARG_list paralist;   /* parameter list */
+enum pj_io_units {
+    PJ_IO_UNITS_CLASSIC = 0,   /* LEFT: Radians     RIGHT: Scaled meters */
+    PJ_IO_UNITS_METERS  = 1,   /* Meters  */
+    PJ_IO_UNITS_RADIANS = 2    /* Radians */
+};
+#ifndef PROJ_H
+typedef struct PJconsts PJ;         /* the PJ object herself */
+#endif
 
-struct PJ_LIST {
-    char             *id;                         /* projection keyword */
-    struct PJconsts  *(*proj)(struct PJconsts*);  /* projection entry point */
-    char * const     *descr;                      /* description text */
+struct PJ_REGION_S {
+    double ll_long;        /* lower left corner coordinates (radians) */
+    double ll_lat;
+    double ur_long;        /* upper right corner coordinates (radians) */
+    double ur_lat;
 };
 
-/* Merging this into the PJ_LIST infrastructure is tempting, but may imply ABI breakage. Perhaps at next major version? */
+
+/* base projection data structure */
+struct PJconsts {
+
+    /*************************************************************************************
+
+                            G E N E R A L   C O N T E X T
+
+    **************************************************************************************
+
+	    Need some description here - especially about the thread context...
+
+    **************************************************************************************/
+
+    projCtx_t *ctx;
+    const char *descr;             /* From pj_list.h or individual PJ_*.c file */
+    paralist *params;              /* Parameter list */
+    struct pj_opaque *opaque;      /* Projection specific parameters, Defined in PJ_*.c */
+
+
+    /*************************************************************************************
+
+                          F U N C T I O N    P O I N T E R S
+
+    **************************************************************************************
+
+	    For projection xxx, these are pointers to functions in the corresponding
+		PJ_xxx.c file.
+
+        pj_init() delegates the setup of these to pj_projection_specific_setup_xxx(),
+		a name which is currently hidden behind the magic curtain of the PROJECTION
+		macro.
+
+		As the PROJ.4 de-macroization project expands its coverage, this will change,
```

Comment:
I suggest leaving this paragraph out. The odds of this message surviving the intendd de-macroization are quite high and will most likely lead to confusion for future readers of the code.